### ISSUE TITLE
Using squash and merge for pull requests

### ISSUE THREAD

**Issue by rfalgout** [hypre-space/hypre#188]
Hello @hypre-space/developers and @jbschroder .  I recently learned of GitHub's *Squash and merge* feature for merging pull requests, and I think we should start using this in the hypre and xbraid projects.  This approach leads to cleaner commit histories on the master branch without losing the detailed commit history on the feature branch and in the pull request.  For more information on this feature, see [About pull request merges](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-request-merges).

For an example of a pull request that was merged in this way, see [this recent one for XBraid](https://github.com/XBraid/xbraid/pull/50).  Notice that the feature branch was deleted and can still be restored as usual.  Also notice the single merge commit [here](https://github.com/XBraid/xbraid/commit/033e1c881d6d27ce02119a8741318f93fe003649).

Below is a section of our [Git Workflow](../wiki/Git-Workflow) Wiki page that I've modified to represent the proposed new process.  Take a look and let me know what you think.  Note that there is currently no way to make this the default merge approach, but it is possible to make it the only option.  When we merge two branches together, we should continue to use the `--no-ff` approach we have used to date, and usually we do that from the command line so maybe this would be okay.  I'd like to hear your comments here as well.

Thanks!

### Create a pull request

Create a pull request to finalize changes before merging into the master branch.

Before merging, it is wise to first update the feature branch (as outlined above) if it is behind the master, then run additional tests (even if the code itself has no conflicts, behavior changes are possible).

Merge the feature into the master branch by selecting *Squash and merge* from the drop-down list on the *Merge pull request* button near the bottom of the pull request.  This will create a single new commit on the master branch for the merge, so please take the time to write an informative commit message.  See our [Git Workflow](../wiki/Git-Workflow) Wiki page for pointers.

Delete the feature branch after the merge.  The easiest way to do this is by clicking a button from within the pull request.  Alternatively, the branch can be deleted from the branch listing page.  GitHub will automatically provide a *Restore branch* button in the closed pull request.  It is a good idea to delete your local branch as well.

    git branch -d foo

**Comment by jbschroder** [comment]
Hi All,

I like the pull request workflow.  Thank you Rob!  I also like the idea of
making this the only option for merges (at least for XBraid).  I can't
think of a case where you wouldn't want to do a squash-merge, but there
probably is some edge case.

Cheers,

Jacob

**Comment by ulrikeyang** [comment]
Hi all,
this sounds good to me. I like that it ends up with overall less stuff, but keeps the important info.
Cheers
Ulrike

**Comment by victorapm** [comment]
Hi all,

I like the idea of squash and merge. I heard about many other groups at the lab using it. I guess we only need to have in mind some of its limitations. For example, it might create some difficulties when trying to merge a branch of a branch into master after the first branch has already been squashed into master (see [this link](https://stackoverflow.com/questions/22593087/merging-a-branch-of-a-branch-after-first-branch-is-squashed-when-merged-to-maste)). But yes, I agree with the idea :)

**Comment by oseikuffuor1** [comment]
Hi all,
this sounds like a good approach to me as well. Thanks @rfalgout. I thought this [article](https://blog.dnsimple.com/2019/01/two-years-of-squash-merge/) had some interesting points. @victorapm's corner case is also an interesting one.

**Comment by rfalgout** [comment]
Thanks for pointing out the corner case @victorapm .  I played around with a private repository just now to see if this would be an issue for us, and as long as we follow our usual practices, we should be fine.  Consider your corner case: branch2 is created from branch1, then both branches are edited separately, then branch1 is squash-merged into master.  The thing to do with the branch2 is to merge in the changes from branch1 so that the two branches are again in sync with respect to master (but unlike the master, branch2 retains the entire commit log).  This should be done with the `--no-ff` option we normally use.  Any additional merges of master into the branch2 should not create unexpected conflicts (I didn't have issues with that anyway), and branch2 can be squash-merged into master with no problem either.  Basically, we should generally deal with branches in the way we normally have, but do squash-merges to master to keep things tidy and easier to follow.

One thing to note with squash-merge is that we should plan to always delete the feature branch once it is merged.  This was already a practice we were trying to follow.  It is also good practice to make sure that the feature branch is not behind the master before it is merged, which is something we already try to do now as well.

Another thing to note is that all of the commit history is retained, even from the command line.  You won't see the details by default, but you will if you use `git log --all` or the longer graphical command mentioned in our Git Workflow Wiki.  Of course, even more detail is retained in the github pull request history.

**Comment by rfalgout** [comment]
Nice article pointer @oseikuffuor1 .  Thanks!

**Comment by rfalgout** [comment]
Unless there are last minute objections, I will go ahead and update our Git Workflow page to adopt this new squash-merge approach.  I'll do the same for the XBraid Wiki @jbschroder .  Thanks!

**Comment by rfalgout** [comment]
I have rewritten the [Git WorkFlow Wiki](https://github.com/hypre-space/hypre/wiki/Git-Workflow) for hypre.  Feedback is welcome.  Thanks!

**Comment by victorapm** [comment]
> I played around with a private repository just now to see if this would be an issue for us

Thanks for letting me know about this, @rfalgout. And thanks @oseikuffuor1 for the article!

**Comment by rfalgout** [comment]
Hi @jbschroder .  I just updated the braid workflow wiki with the squash merge information.
